### DISCUSSION TITLE
React/NextJS and Searchbar

### DISCUSSION THREAD

**Discussion by a95tbirdsc** [vercel/next.js#48971] [Help]
### Summary

I have simple React/NextJS Employee Directory application but I can't seem to figure out how to wire up the SearchBar user input to the Result page.

So, the application works, displays an input field with the Fetch results below that.  That all works.

What I can't figure out is how to get what the user types in the input field to the SearchResults page to show the filtered results.

The SearchText component, which has the input field, has the following:
"use client"
"const [query, setQuery] = useState('');"

input field has: value={query} onChange={(e) => setQuery(e.target.value)}

SearchResults has the following (just a snippet of code):

import SearchText from '../search-text/page.js';

export default async function SearchResults({ query })

async function getEmployees() {
    const res = await fetch('https://xxxx?searchtype=Input&searchtext=${query}');
    const directory = await res.json();
    return directory;
  }

How do I get what is typed into the Input field via the QUERY parameter to be a parameter to the SearchResults component?

**Comment by SamKomesarook** [comment]
If SearchResults is a component, you could pass the args as follows:
`<SearchResults query />`

Could you share the components in more detail?

**Comment by a95tbirdsc** [comment]
From what other people have told me it can't be done.  You can't have a simple search using "onChange" with React/Next JS using the APP folder structure.  You can't pass an argument from a Client Component to a Server Component.  It would be wonderful if you could show me how to do it.

What am I trying to do?  Pretty simple.  Using the APP folder structure.  A simple INPUT field that when a user types (onChange) in information (J..JA..JAME...JAMES) then the API is called with each KEYSTROKE to return a result set based on the user's input.  The Results Section is then updated with that information.

SearchText
```
'use client'

import { useCallback, useState } from 'react'
import utilStyles from '/styles/utils.module.css';
import { faMagnifyingGlass, faPeopleGroup, faBuildingUser, faArrowsRotate } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import SearchResults from '../search-results/page.js';

export default function SearchText() {
  const [query, setQuery] = useState('');
  
  const onChange = useCallback((event) => {
    const query = event.target.value;
    setQuery(query);
  }, []);

  console.log({query});
  return (
      
    <>
      
      <form id='EmployeeSearchText'>
        <div className="container-fluid p-3">
          <div id='Row1' className={`row ${utilStyles.divPadding}`}>
            <div id='Row1Column1' className="col">
              <div id='SearchText-container' className={utilStyles.searchBar}>
                <FontAwesomeIcon icon={faMagnifyingGlass} className={utilStyles.icon} />
                <input id='SearchText' type="text" onChange={onChange} className={`form-control ${utilStyles.inputField} ${utilStyles.inputPadding}`} name="SearchText" value={query} placeholder="Search Last Name; First Name; Campus; Deparment....." />
              </div>
            </div>
          </div>
          <div className='text-center'>
            <div id='Row2' className={`row ${utilStyles.divPadding}`}>
              <div id='Row2Column1' className="col">
                <div id='Employees-container'>
                  <button id='EmployeeListing' className={utilStyles.btnYellow}>Employees&nbsp;&nbsp;<FontAwesomeIcon icon={faPeopleGroup} /></button>
                </div>
              </div>
              <div id='Row2Column2' className="col">
                <div id='DepartmentListing-container'>
                  <button id='DepartmentListing' className={utilStyles.btnYellow} type="button">Department Listing&nbsp;&nbsp;<FontAwesomeIcon icon={faBuildingUser} /></button>
                </div>
              </div>
              <div id='Row2Column3' className="col">
                <div id='SearchTextReset-container'>
                  <button id='SearchTextReset' className={utilStyles.btnGrey} type="button">Reset&nbsp;&nbsp;<FontAwesomeIcon icon={faArrowsRotate} /></button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>

      
      
    </>
  
  )
}

```

SearchResults

```
import utilStyles from '/styles/utils.module.css';
import { faPhone, faEnvelope } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export default async function SearchResults( {query} ) {
  console.log("Results " + {query});
  async function getEmployees() {
    const res = await fetch('https://xxx?searchinput=${query}');
    
    const directory = await res.json();
    return directory;
  }
  
  console.log({query});
  const employees = await getEmployees();

  const emp = employees.map((employee, index, array) => {
    
    return (
        
      <div className="col col-12 col-sm-6 col-xl-4 mb-4">
        <div className="col-12 mb-4">
          <div className={`card ${utilStyles.cardBody}`}>
            <div className="card-body">
              <div className={`${utilStyles.ribbonLight} ${utilStyles.ribbonBottomRight}`}><span>{employee.CAMPUS}</span></div>
              <h3 className="card-title card-name">{employee.NAME}</h3>
              <h5 className={`${utilStyles.cardSubtitle} ${utilStyles.cardJobtitleNew} mb-2`}>{employee.JOBTITLE}</h5>
              <p className={`${utilStyles.cardText} ${utilStyles.cardDepartment}`}>{employee.DEPARTMENT}</p>
              <div>
                <p className={utilStyles.cardTextBottom}><FontAwesomeIcon icon={faPhone} className={utilStyles.cardTextIcons} /><span className={utilStyles.cardPhoneEmail}>W:</span>{employee.PHONE}</p>
                <p className={utilStyles.cardTextBottom}><FontAwesomeIcon icon={faEnvelope} className={utilStyles.cardTextIcons} /><span className={utilStyles.cardPhoneEmail}>W:</span></p>
              </div>
            </div>
          </div>
        </div>
      </div>
    
    )
      
  });
  
  return (
    
    <div className="container">
      <div className="row">{emp}</div>
    </div>
       
  );
}

```

**Reply by SamKomesarook** [reply]
If you intend for the `SearchResults` component to be a server one, then you're right, it isn't possible, as the contained data will be dynamic based on the user input...

Is there a requirement for the `SearchResults` component to be a server one? Why?

**Reply by a95tbirdsc** [reply]
Not really but I thought the whole point of the APP folder was to keep data fetching on the server.

I mean, has no one at Vercel (Next JS) thought of that?  Seems pretty basic.  Most times fetching data is dynamic, yet, you can't even have a simple input field that allows a user to filter data.  Or what about radio buttons, check boxes, drop downs or anything else interactive.

**Reply by SamKomesarook** [reply]
You're right, ideally data that needs to be fetched should be fetched at build time on the server, however as far as I know, anything that requires user-input will happen within a client-component. Note the checklist in the [docs](https://nextjs.org/docs/getting-started/react-essentials#when-to-use-server-and-client-components) where interactivity needs to be handled by client components.

**Reply by a95tbirdsc** [reply]
That's crazy.  You have a framework that won't even allow you to do what it's suppose to do.  Fetch data as a Server Component but it can't be interactive.  That has to be one of the dumbest things I have every heard.

I understand that interactivity is handled by the Client Component but you would think there would be some way to pass that information to a Server Component to get whatever data is needed.  I believe it can be done with submitting a form and clicking a button but what do you do if your elements have "onChange" so it's more dynamic.  Nothing I guess.  You can't do it.

**Reply by DevAtHeart** [reply]
As a workaround, you can write the query into the searchParams in the URL. But then we have our state in the url. I don't get it either.

Like the example here: https://nextjs.org/docs/getting-started/react-essentials#patterns

What does that awesome searchbar even do? It's hilarious.
